### PR TITLE
decompose cluster creation into reusable modules

### PR DESCRIPTION
**Pull request by houqp**
@pgavlin , this PR aims to finish off your previous WIP commit. Once it's in usable state, i am planning to catch up with master. Then lastly my patches from https://github.com/pulumi/pulumi-eks/pull/22.

Let me know if you prefer to approach this in a different direction.

### CODE REVIEW

File: nodejs/eks/workerpool.ts [hunk 1]
```diff
@@ -114,32 +124,77 @@ export class WorkerPool extends pulumi.ComponentResource {
     constructor(name: string, args: WorkerPoolOptions, opts?: pulumi.ComponentResourceOptions) {
         super("eks:index:WorkerPool", name, args, opts);
 
-        const pool = createWorkerPool(name, args, this, this.getProvider("kubernetes:core:v1/ConfigMap"));
+        const pool = createWorkerPool(name, args, this);
         this.instanceRole = pool.instanceRole;
         this.nodeSecurityGroup = pool.nodeSecurityGroup;
-
-        this.registerOutputs();
```
Comment (pgavlin) [inline nodejs/eks/workerpool.ts]: This call is required in order to inform the CLI's display routine that the component has finished creating. Let's keep it as-is.

File: nodejs/eks/workerpool.ts [hunk 2]
```diff
@@ -49,11 +44,26 @@ export interface WorkerPoolOptions {
      */
     cluster: aws.eks.Cluster;
 
+    /**
+     * k8s provider for this worker pool
+     */
+    provider: k8s.Provider;
```
Comment (pgavlin) [inline nodejs/eks/workerpool.ts]: `provider` should not be a part of `WorkerPoolOptions`: it is already part of `pulumi.ComponentResourceOptions`, which is passed to the `WorkerPool` constructor. I'd rather we leave it as a separate parameter on `createWorkerPool`.

File: nodejs/eks/workerpool.ts [hunk 3]
```diff
@@ -150,60 +205,24 @@ export function createWorkerPool(name: string, args: WorkerPoolOptions, parent:
...
     const eksClusterIngressRule = new aws.ec2.SecurityGroupRule(`${name}-eksClusterIngressRule`, {
         description: "Allow pods to communicate with the cluster API Server",
         type: "ingress",
         fromPort: 443,
         toPort: 443,
         protocol: "tcp",
-        securityGroupId: args.eksClusterSecurityGroup.id,
-        sourceSecurityGroupId: nodeSecurityGroup.id,
+        securityGroupId: args.clusterSecurityGroup.id,
+        sourceSecurityGroupId: this.nodeSecurityGroup.id,
```
Comment (pgavlin) [inline nodejs/eks/workerpool.ts]: I think this should just be `nodeSecurityGroup` (same on line 224).

File: nodejs/eks/workerpool.ts (hunk 2 above)
Comment (houqp) [reply nodejs/eks/workerpool.ts]: I see, I did not know that. What's the right way to get the k8s provider from within the constructor through ComponentResourceOptions? The original `this.getProvider("kubernetes:core:v1/ConfigMap")` code results in type error.

Comment (pgavlin) [reply nodejs/eks/workerpool.ts]: What was the type error? That method should exist.

Comment (pgavlin) [reply nodejs/eks/workerpool.ts]: If the type error was complaining about the value possibly being `undefined`, you can do something like:

```
const k8sProvider = this.getProvider("kubernetes:core:v1/ConfigMap");
if (k8sProvider === undefined) {
    throw new RunError("a 'kubernetes' provider must be specified for a 'WorkerPool'");
}
```

Comment (houqp) [reply nodejs/eks/workerpool.ts]: This is exactly what I needed, thanks a lot for the tip :)

**Comment by houqp** [comment]
@pgavlin ready for another round of review. Do you prefer me to cherry pick master commits into this PR or use a separate PR for master sync?

**Comment by pgavlin** [comment]
This looks good to me. Let's use a separate PR into this branch to bring it up-to-date against master.

**Pull request merged**